python/prairielearn.py

```diff
@@ -96,13 +96,19 @@ def all_partial_scores_correct(data: QuestionData) -> bool:
     )
 
 
-def to_json(v):
+def to_json(v, *, np_encoding=1):
```

Comment:
What's the migration path to the new encoding? Will we update all elements and example questions to use it? How do we communicate this change to courses? Is there a plan for deprecating and removing the legacy encoding entirely someday?

Is it strictly necessary to put this behind a flag? AFAICT the only case where this would cause issues is if someone was using `to_json` to store data but not using `from_json` to retrieve it, which feels like a question bug that we shouldn't feel the need to support.

If we decide we definitely need to flag this, the name `np_encoding` feels weird to me. I'd be inclined to make this either `np_encoding_version=1` or `legacy_np_encoding=True`.

Reply:
Just to clarify, in a previous comment, I mistakenly said that numpy scalars were being auto-encoded as their corresponding Python types. No idea why I observed this particular behavior for a while, but PL just throws an error in this case as was noted in the original issue.

> What's the migration path to the new encoding? Will we update all elements and example questions to use it?

This encoding type is mainly for use by instructors writing custom questions (this seems to be the use case mentioned in the original issue #1366), so I don't think that there really needs to be a migration for core PL code. In particular, using this for the `pl-number-input` would change the way that some internal data is stored, which in turn would break custom grader code (including custom grader code used in internal test questions). So I think the best way to introduce this is by having it be opt-in for those that want it. 

> How do we communicate this change to courses?

Since this change is opt-in, I think we can communicate it through a news item of some type. There has been some discussion about a news item for the recent Python changes, and I'd be happy to whip up a combined one once we merge in some more of these instructor-facing changes.

> Is there a plan for deprecating and removing the legacy encoding entirely someday?

So the current legacy encoding is to do nothing (not sure why I thought that numpy types were being auto-encoded as their Python counterparts, testing again reveals that it just raises an error, as in the original issue), so I don't think there's much of a point to deprecating this. Alternatively, we could just make this particular encoding a different function altogether, but I think that could cause more confusion than it's worth.

> Is it strictly necessary to put this behind a flag? AFAICT the only case where this would cause issues is if someone was using `to_json` to store data but not using `from_json` to retrieve it, which feels like a question bug that we shouldn't feel the need to support.

As discussed above, there are cases where people (including core-PL code) run `to_json` and expect it to do nothing (don't know why these were the semantics that were decided upon, but that's how it works), and so the flag prevents breaking this in those cases.

> If we decide we definitely need to flag this, the name `np_encoding` feels weird to me. I'd be inclined to make this either `np_encoding_version=1` or `legacy_np_encoding=True`.

I think I'm leaning towards `np_encoding_version`, as @coatless has a preference against using a boolean flag.

Reply:
Ok so I figured it out. Weirdly, the `float64` numpy type is json serializable (but get its type erased when decoding), and the other types are not. See [here](https://stackoverflow.com/questions/44459168/why-are-some-numpy-datatypes-json-serializable-and-others-not) for details. Either way, this means that the new conversion is technically backwards incompatible for some cases, so I think should be behind a flag (since some code relies on the above default behavior, including code in the number input element).